src/util/threadpool.h

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2024-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_UTIL_THREADPOOL_H
+#define BITCOIN_UTIL_THREADPOOL_H
+
+#include <sync.h>
+#include <util/string.h>
+#include <util/thread.h>
+#include <util/threadinterrupt.h>
+
+#include <algorithm>
+#include <atomic>
+#include <condition_variable>
+#include <cstddef>
+#include <functional>
+#include <future>
+#include <memory>
+#include <stdexcept>
+#include <utility>
+#include <queue>
+#include <thread>
+#include <vector>
+
+class ThreadPool {
+
+private:
+    std::string m_name;
+    Mutex m_mutex;
+    std::queue<std::function<void()>> m_work_queue GUARDED_BY(m_mutex);
+    std::condition_variable m_cv;
+    std::atomic<bool> m_interrupt{false};
+    std::vector<std::thread> m_workers;
+
+    void WorkerThread() EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        WAIT_LOCK(m_mutex, wait_lock);
+        while (!m_interrupt.load()) {
+            std::function<void()> task;
+            {
+                // Wait for the task or until the stop flag is set
+                m_cv.wait(wait_lock,[&]() EXCLUSIVE_LOCKS_REQUIRED(m_mutex) { return m_interrupt.load() || !m_work_queue.empty(); });
+
+                // If stopped and no work left, exit worker
+                if (m_interrupt && m_work_queue.empty()) {
+                    return;
+                }
+
+                task = std::move(m_work_queue.front());
+                m_work_queue.pop();
+            }
+
+            {
+                // Execute the task without the lock
+                REVERSE_LOCK(wait_lock, m_mutex);
+                task();
+            }
+        }
+    }
+
+public:
+    ThreadPool(const std::string& name) : m_name(name) {}
+
+    ~ThreadPool()
+    {
+        Stop(); // In case it hasn't been stopped.
+    }
+
+    void Start(int num_workers)
+    {
+        if (!m_workers.empty()) throw std::runtime_error("Thread pool already started");
+        m_interrupt.store(false); // Reset
+
+        // Create workers
+        for (int i = 0; i < num_workers; i++) {
+            m_workers.emplace_back(&util::TraceThread, m_name + "_threadpool_worker_" + util::ToString(i), [this] { WorkerThread(); });
```

Comment:
This is bike shedding now so ok to ignore -- but when we set thread names on Linux they are truncated to 15 characters. So I dunno I guess "thread" and "worker" are redundant in the name of a thread?

https://github.com/bitcoin/bitcoin/blob/2f410ad78c767e37f083d03114f6661b73647af3/src/util/threadnames.cpp#L22-L27

<img width="692" height="317" alt="image" src="https://github.com/user-attachments/assets/8819caee-2731-462c-a914-c595a8599000" />